File: LeetCode_Algorithm/323(Number of Connected Components in an Undirected Graph)/323.py
'''
LeetCode 323 : Number of Connected Components in an Undirected Graph
무방향 그래프가 존재할 때 n노드가 존재할 때 연결된 그룹을 구하라.

# sol1 100ms 83% 15.7MB 50%
dfs로 순회하며 저장
'''
class Solution:
    def countComponents(self, n: int, edges: list[list[int]]) -> int:
        from collections import defaultdict
        graph = defaultdict(list)
        for [i, j] in edges:
            graph[i].append(j)
            graph[j].append(i)
        seen = set(); ans = 0
        for i in range(n):
            if i not in seen:
                stack = [i]
                while stack:
                    node = stack.pop()
                    if node not in seen:
                        seen.add(node)
                    else:
                        continue
                    for n in graph[node]:
                        stack.append(n)
                ans += 1
        return ans
                    

File: LeetCode_Algorithm/323(Number of Connected Components in an Undirected Graph)/323.js
var countComponents = function(n, edges) {
  // sol1 89ms 45% 48.1MB 8%
  let graph = new Map()
  for (let i = 0; i < n; i++) {
    graph.set(i, [])
  }
  for (let [i, j] of edges) {
    let val = graph.get(i); val.push(j)
    graph.set(i, val)
    val = graph.get(j); val.push(i)
    graph.set(j, val)
  }
  let seen = new Set(); let ans = 0
  for (let i = 0 ; i < n ; i++ ){
    if (seen.has(i))
      continue
    let stack = [i]
    while (stack.length) {
      let node = stack.pop()
      if (seen.has(node))
        continue
      seen.add(node)
      let iter = graph.get(node)
      for (let i of iter) {
        stack.push(i)
      }
    }
    ans += 1
  }
  return ans
};

console.log(countComponents(5, [[0,1],[1,2],[2,3],[3,4]]))